bin/internal/prepare-workspace.sh

```diff
@@ -0,0 +1,121 @@
+#!/bin/sh
+
+################################################################################
+# This program and the accompanying materials are made available under the terms of the
+# Eclipse Public License v2.0 which accompanies this distribution, and is available at
+# https://www.eclipse.org/legal/epl-v20.html
+#
+# SPDX-License-Identifier: EPL-2.0
+#
+# Copyright IBM Corporation 2020
+################################################################################
+
+################################################################################
+# This script will run component `validate` and `configure` step if they are defined.
+#
+# This script take these parameters
+# - c:    INSTANCE_DIR
+# - t:    a list of component IDs separated by comma
+#
+# For example:
+# $ bin/internal/prepare-workspace.sh \
+#        -c "/path/to/my/zowe/instance" \
+#        -t "discovery,explorer-jes,jobs"
+################################################################################
+
+# if the user passes INSTANCE_DIR from command line parameter "-c"
+while getopts "c:t:" opt; do
+  case $opt in
+    c) INSTANCE_DIR=$OPTARG;;
+    t) LAUNCH_COMPONENTS=$OPTARG;;
+    \?)
+      echo "Invalid option: -$OPTARG" >&2
+      exit 1
+      ;;
+  esac
+done
+
+########################################################
+# prepare environment variables
+export ROOT_DIR=$(cd $(dirname $0)/../../;pwd)
+. ${ROOT_DIR}/bin/internal/prepare-environment.sh -c "${INSTANCE_DIR}"
+
+########################################################
+# Validate component properties if script exists
+ERRORS_FOUND=0
+for component_id in $(echo $LAUNCH_COMPONENTS | sed "s/,/ /g")
+do
+  component_dir=$(find_component_directory "${component_id}")
+  # backward compatible purpose, some may expect this variable to be component lifecycle directory
+  export LAUNCH_COMPONENT="${component_dir}/bin"
+  # FIXME: change here to read manifest `commands.validate` entry
+  VALIDATE_SCRIPT=${component_dir}/bin/validate.sh
+  if [ ! -z "${component_dir}" -a -x "${VALIDATE_SCRIPT}" ]; then
+    . ${VALIDATE_SCRIPT}
+    retval=$?
+    let "ERRORS_FOUND=$ERRORS_FOUND+$retval"
+  fi
+done
+# exit if there are errors found
+checkForErrorsFound
+
+########################################################
+# Prepare workspace directory
+mkdir -p ${WORKSPACE_DIR}
+# Make accessible to group so owning user can edit?
+chmod -R 771 ${WORKSPACE_DIR}
+
+# Copy manifest into WORKSPACE_DIR so we know the version for support enquiries/migration
+cp ${ROOT_DIR}/manifest.json ${WORKSPACE_DIR}
+
+# Keep config dir for zss within permissions it accepts
+# FIXME: this should be moved to zlux/bin/configure.sh
```

Comment:
Ideally we want this removed entirely as it stops uses from being able to delete the instance directory and can cause errors on upgrade

Reply:
added your comments into source code